### PR TITLE
Small optimization + cleaning code

### PR DESCRIPTION
**Pull request by wagenaartje**
A summary of the changes I made:

- Do not use `for(var i in array)`! For arrays, always use `for(var i = 0; i < array.length; i++)`. Why? Have a look yourself: https://jsperf.com/var-i-in-array-vs-var-i-is-0/1. For me, thats a 97% speed difference! So basically, I changed this in the code as much as I can. 
- Instead of iterating through the list to make a 'copied' list of network.layers.hidden, and then reversing it - I simply iterate backwards over the hidden layers to propagate: https://jsperf.com/iterate-plus-reverse-vs-reverse-reverse. Thats a 83% speed difference for me.
- Using `set` could cause network.layers.hidden to be `undefined`, so instead of doing `this.layers = layers`, it now works like:

```javascript
    this.layers = {
      input:  layers.input || null,
      hidden: layers.hidden || [],
      output: layers.output || null
    };
```
- `--require` does not come installed when you use `npm install`. And it is an unnecessary dependency anyways: just correctly using `_inject-dist/src.js` in `test/` for mocha testing is far more practical. Also removed `mocha.opts` because it served no use.
- Added some brackets for readability (it's just personal style, so I let it pretty much unchanged mostly)
- There was difficult computation for `Neuron.squash.TANH` (maybe for a reason?), but I changed it to just use the built-in method `Math.tanh()`
- Neuron squases no longer call themselves,  this is inefficient.
- `network.label` and `node.label` properties removed. They weren't used anymore and served no purpose. If a user wants to set a custom label, he/she can just do `instance.label = `. 
- `Perceptron`, `LSTM`  and `Liquid` all used `this.trainer = new Trainer(this)`. Why is this not stated in the documentation? The documentation even encourages:

```javascript
var myNetwork = new Architect.Perceptron(2, 2, 1)
var trainer = new Trainer(myNetwork)
```
So I removed the Architect constructors to automatically create a `Trainer` for themselves (except for Hopfield). This was only used in mocha testing anyways, which I updated.
- Actually made `synaptic.min.js` minified! Using uglifyJS webpack plugin. So I got these warnings:
warnings but I only changed the unused variable because I wasn't sure what the reasoning was behind the static if statements.
- Some other things

If you feel like I should change something about this, feel free to let me know! I might do some more optimization in the future.

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -9,9 +10,15 @@ module.exports = {
...
     path: 'dist',
     filename: '[name].js',
+    library: 'synaptic',
+    libraryTarget: 'umd'
   },
   plugins: [
     new webpack.NoErrorsPlugin(),
-    new webpack.BannerPlugin(license())
+    new webpack.BannerPlugin(license()),
+    new webpack.optimize.UglifyJsPlugin({
```
Comment (cazala) [inline webpack.config.js]: The reason we were not minifying was because that was breaking the `trainAsync` methods. You can test it if you add an `index.html` file in the repo root directory with the following:

```html
<script src="./dist/synaptic.min.js"></script>
<script>
  var net = new synaptic.Architect.Perceptron(2, 4, 1);
  var trainer = new synaptic.Trainer(net);
  trainer.trainAsync([
    {
      input: [0, 0],
      output: [0]
    },
    {
      input: [0, 1],
      output: [1]
    },
    {
      input: [1, 0],
      output: [1]
    },
    {
      input: [1, 1],
      output: [0]
    }
  ], { log: 1 });

</script>
```

You will see an error in the console, and if you change the `src` to `./dist/synaptic.js` it will work again.

This is something that won't happen anymore in the next version, but I don't think it will get fixed on this one, so for now we should remove this plugin from the build.

**Comment by cazala** [comment]
Awesome @wagenaartje :D thanks for the PR, I just left one change request regarding the minification, everything else looks great 👍

**Comment by wagenaartje** [comment]
Oops, didn't know about that. I just reverted the uglifyJS in a commit and made another commit to let the TravisCI check rerun (sometimes it just fails). Should be good to go :+1:
